backend/nokia-innovative-project/src/main/java/com/nokia/library/nokiainnovativeproject/services/UserService.java

```diff
@@ -3,51 +3,164 @@
+        Hibernate.initialize(user.getBooks());
+        Hibernate.initialize(user.getAddress());
+        Hibernate.initialize(user.getRoles());
+        return detachAndSecure(user);
     }
 
     public User createUser(UserDTO userDTO) {
+        if(userRepository.countUserByEmail(userDTO.getEmail()) > 0){
+            throw new ValidationException("User with this email already exist!");
+        }
         ModelMapper mapper = new ModelMapper();
         User user = mapper.map(userDTO, User.class);
-        return userRepository.save(persistingRequiredEntities(user, userDTO));
+        user.setPassword(bCryptPasswordEncoder.encode(userDTO.getPassword()));
+        user.setRoles(Arrays.asList(roleRepository.findByRole("ROLE_EMPLOYEE")));
+        user = userRepository.saveAndFlush(persistingRequiredEntities(user, userDTO));
+        user = detachAndSecure(user);
+        user.setRoles(new ArrayList<>());
+        return user;
     }
 
     public User updateUser(Long id, UserDTO userDTO) {
         User user = userRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("user"));
+        if(user.getEmail() != userDTO.getEmail()) {
+            throw new ValidationException("You can't change email!");
+        }
         user.setFirstName(userDTO.getFirstName());
         user.setLastName(userDTO.getLastName());
         user.setEmail(userDTO.getEmail());
-        user.setAddress(userDTO.getAddress());
-        return userRepository.save(user);
+        user.setPassword(bCryptPasswordEncoder.encode(userDTO.getPassword()));
+        user = userRepository.saveAndFlush(persistingRequiredEntities(user, userDTO));
+        user = detachAndSecure(user);
+        user.setRoles(new ArrayList<>());
+        return user;
+    }
+
+    public User assignAdminRoleToUser(Long id) {
+        User user = userRepository.findById(id).orElseThrow(
+                ()-> new ResourceNotFoundException("user"));
+        if(user.getAddress() == null) {
+            throw new ValidationException("User has no address. First, assign the address to the user.");
+        }
+        List<Role> userRoles = user.getRoles();
+        for(Role role : userRoles) {
+            if(role.getRole().equals("ROLE_ADMIN"))
+                return user;
+        }
+        Role role = roleRepository.findByRole("ROLE_ADMIN");
+        userRoles.add(role);
+        user.setRoles(userRoles);
+        user = userRepository.saveAndFlush(user);
+        return detachAndSecure(user);
+    }
+
+    public User takeAdminRoleFromUser(Long id) {
+        if(userRepository.countUserByRole("ROLE_ADMIN") <= 1) {
+            throw new ValidationException("You can't delete the last admin from the database");
+        }
+        User user = userRepository.findById(id).orElseThrow(
+                ()-> new ResourceNotFoundException("user"));
+        Hibernate.initialize(user.getBooks());
+        Hibernate.initialize(user.getRoles());
+        Hibernate.initialize(user.getAddress());
```

Comment:
I use `@EntityGraph` in our repository interface, maybe it would help you with this initialization?
```
	@Override
	@EntityGraph(attributePaths = {
			"books",
			"roles",
			"address",
	})
	List<User> findAll();
``` 
it loads those dependencies with 1 sql query and joins